tests/tests_data/streaming/test_dataset.py

```diff
@@ -556,66 +593,101 @@ def test_resumable_dataset_single_worker(tmpdir):
+    dataset = EmulateS3StreamingDataset(
+        input_dir=RemoteDir(cache_dir, data_dir), item_loader=TokensLoader(block_size), shuffle=True
+    )
 
     dataset.current_epoch = 1
 
-    assert dataset.state_dict() == {}
+    # assert dataset.state_dict() == {}
 
-    dataloader = DataLoader(dataset, num_workers=1, batch_size=2, prefetch_factor=1)
+    dataloader = DataLoader(dataset, num_workers=2, batch_size=2, prefetch_factor=1)
 
     dataloader_iter = iter(dataloader)
 
     _ = next(dataloader_iter)
-    state_dict_0 = dataset.state_dict()
 
     sleep(0.1)
 
-    assert state_dict_0["0"]["chunk_index"] == 0
+    state_dict_0 = dataset.state_dict()
+
+    assert sorted(state_dict_0.keys()) == ["0", "1"]
+
+    assert state_dict_0["0"]["chunk_index"] == 1
+    assert state_dict_0["0"]["global_index"] == 2
     assert state_dict_0["0"]["index"] == 0
 
-    checkpoint_dir = os.path.join(tmpdir, "checkpoints")
-    assert os.listdir(checkpoint_dir) == ["0"]
+    assert state_dict_0["1"]["chunk_index"] == 0
+    assert state_dict_0["1"]["global_index"] == 0
+    assert state_dict_0["1"]["index"] == 0
+
     _ = next(dataloader_iter)
 
     sleep(0.1)
 
     state_dict_1 = dataset.state_dict()
-    assert state_dict_1["0"]["chunk_index"] == 2
+
+    assert state_dict_1["0"]["chunk_index"] == 1
+    assert state_dict_1["0"]["global_index"] == 2
     assert state_dict_1["0"]["index"] == 0
 
+    assert state_dict_1["1"]["chunk_index"] == 1
+    assert state_dict_1["1"]["global_index"] == 2
+    assert state_dict_1["1"]["index"] == 0
+
     batch_2 = next(dataloader_iter)
 
     sleep(0.1)
 
     state_dict_2 = dataset.state_dict()
-    assert state_dict_2["0"]["chunk_index"] == 3
+
+    assert state_dict_2["0"]["chunk_index"] == 2
+    assert state_dict_2["0"]["global_index"] == 4
     assert state_dict_2["0"]["index"] == 0
 
-    dataset = StreamingDataset(input_dir=str(tmpdir), item_loader=TokensLoader(block_size), shuffle=True)
+    assert state_dict_2["1"]["chunk_index"] == 1
+    assert state_dict_2["1"]["global_index"] == 2
+    assert state_dict_2["1"]["index"] == 0
+
+    dataset = EmulateS3StreamingDataset(
+        input_dir=RemoteDir(cache_dir, data_dir), item_loader=TokensLoader(block_size), shuffle=True
+    )
+
     dataset.load_state_dict(state_dict_1)
-    dataloader = DataLoader(dataset, num_workers=1, batch_size=2, prefetch_factor=1)
+    dataloader = DataLoader(dataset, num_workers=2, batch_size=2, prefetch_factor=1)
 
     dataloader_iter = iter(dataloader)
     batch_0_restart = next(dataloader_iter)
 
     sleep(0.1)
 
     state_dict_2 = dataset.state_dict()
-    assert state_dict_2["0"]["chunk_index"] == 3
+
+    assert state_dict_2["0"]["chunk_index"] == 2
+    assert state_dict_2["0"]["global_index"] == 4
     assert state_dict_2["0"]["index"] == 0
 
+    assert state_dict_2["1"]["chunk_index"] == 1
+    assert state_dict_2["1"]["global_index"] == 2
+    assert state_dict_2["1"]["index"] == 0
```

Comment:
Can you remind me what "index" is and why is it 0 in all cases here?

Reply:
This is the index within a chunk.